File: 2 Assignment 7/Board.java
/*
 * Ryan Lutz
 * 11/6/19
 * M2A7
 */

public class Board {
    private char[][] squares; // Represents the game board

    /**
     * Contructs a new Board with 10 rows and 10 columns
     */
    public Board() {
        squares = new char[10][10];
        for (int i=0; i<squares.length; i++) {
            for (int j=0; j<squares[i].length; j++) {
                squares[i][j] = '-';
            }
        }
    }

    /**
     * ToString method for Board class
     * @return A string representing the current game board
     */
    public String toString() {
        String out = "";
        for (char[] row: squares) { // For each row in squares
            for (char e: row) { // For each column in row
                out += e + " "; // Add the element to out and a space after it
            }
            out += "\n"; // Add a newline char after every row
        }
        return out; // Return the out string
    }
    /**
     * Attempts to add a ship of len at (row, col) on either the horizontal or vertical axis
     * @param row row in which the ship should start
     * @param col column in which the ship should start
     * @param len length of the ship
     * @param horizontal if the ship should be horizontal (true) or vertical (false)
     * @return true if the ship was added
     * <li>false if the ship could not be added</li>
     */
    public boolean addShip(int row, int col, int len, boolean horizontal) {
        int f = 0; // Flag var
        if (horizontal) { // Ship wants to be horizontal
            for (int i=0; i<len; i++) { // For each index from column i to i+len
                try {
                    if (squares[row][col+i] != '-') { // If a square isn't empty
                        f = 1; // Raise the flag
                    }
                } catch (IndexOutOfBoundsException e) { // If index goes out of bounds
                    f = 1; // Raise the flag
                }
            }
            if (f == 0) { // If flag has not been set
                for (int i=0; i<len; i++) { // Set each square that the user wants the ship to be in to 'b'
                    squares[row][col+i] = 'b';
                }
                return true;
            } else { // If flag has been set
                return false;
            }
        } else { // Ship wants to be vertical
            for (int i=0; i<len; i++) { // For each index from row i to i+len
                try {
                    if (squares[row+i][col] != '-') { // If a square isn't empty
                        f = 1; // Raise the flag
                    }
                } catch (IndexOutOfBoundsException e) { // If index goes out of bounds
                    f = 1; // Raise the flag
                }
            }
            if (f == 0) { // If flag has not been set
                for (int i=0; i<len; i++) { // Set each square that the user wants the ship to be in to 'b'
                    squares[row+i][col] = 'b';
                }
                return true;
            } else { // If flag has been set
                return false;
            }
        }
    }
    /**
     * Attempts a shot at the specified row and column
     * @param row row of the shot
     * @param col column of the shot
     * @return an int representing if the shot hit, missed, was already shot, or was out of bounds
     * <ul><li>-1 if the shot was out of bounds</li>
     * <li>0 if the shot missed</li>
     * <li>1 if the shot hit</li>
     * <li>2 if the shot had already been taken</li></ul>
     */
    public int shoot(int row, int col) {
        try {
            if (squares[row][col] == '-') { // Shot missed
                squares[row][col] = 'm';
                return 0;
            } else if (squares[row][col] == 'b') { // Ship hit
                squares[row][col] = 'x';
                return 1;
            } else { // Square already guessed
                return 2;
            }
        } catch (IndexOutOfBoundsException e) { // Out of bounds
            return -1;
        }
    }
    /**
     * Checks if there is still a ship left on the board
     * @return true if there is no ship left
     * <li>false if a ship is still left</li>
     */
    public boolean gameOver() {
        for (char[] row: squares) { // For each row in squares
            for (char e: row) { // For each column in row
                if (e == 'b') { // If that element equals 'b'
                    return false;
                }
            }
        }
        return true; // If the loops have ended and no 'b' was found, then game over
    }
    /**
     * Checks if a ship of length len is found either vertically or horizontally on the board
     * @param len length of the ship being searched for
     * @return true if the ship was found
     * <li>false otherwise</li>
     */
    public boolean foundShip(int len) {
        char[][] temp = new char[squares.length][squares[0].length]; // Copy the squares array into a temp array
        for (int i=0; i<squares.length; i++) {
            for (int j=0; j<squares[i].length; j++) {
                temp[i][j] = squares[i][j];
            }
        }
        for (int i=0; i<temp.length; i++) { // For each row in temp
            for (int j=0; j<temp[i].length; j++) { // For each element in row
                if (temp[i][j] == 'b') { // If the element is b
                    int h = 1; // Declare vars for tracking vert and horiz ship length
                    int v = 1; // Initialized to 1 because a square of ship has already been found
                    temp[i][j] = '-'; // Remove the val in the temp array so that it won't trigger another event
                    for (int m=1; m<squares.length-i && m<squares[i].length-j; m++) { // While m is in bounds
                        if (temp[i][j+m] == 'b') { // If any of the squares to the right of [i][j] are also 'b'
                            h++; // Increment h
                            temp[i][j+m] = '-'; // Remove the val in the temp array so that it won't trigger another event
                        }
                        if (temp[i+m][j] == 'b') { // If any of the squares below [i][j] are also 'b'
                            v++; // Increment v
                            temp[i+m][j] = '-'; // Remove the val in the temp array so that it won't trigger another event
                        }
                    }
                    h = h > v ? h : 0; // Basically decides which direction the ship is facing
                    v = v > h ? v : 0; // (the larger value is preserved, the smaller val is set to 0)
                    if (h == len || v == len) { // If the horiz or vert ship length equals the input length
                        return true;
                    }
                }
            }
        }
        return false; // If no ship of correct length found, return false
    }
}

File: 2 Assignment 7 kt/Board.kt
class Board {
  private var squares:Array<Array<Char>>

  constructor() {
    squares = Array<Array<Char>>(10, {Array<Char>(10, {_ -> '-'})})
  }

  override fun toString():String {
    var out:String = ""
    for (row:Array<Char> in squares) {
      for (e:Char in row) {
        out += e + " "
      }
      out += "\n"
    }
    return out
  }

  fun addShip(row:Int, col:Int, len:Int, horizontal:Boolean):Boolean {
    when (horizontal) {
      true -> {
        if (col+len > 10) {
          return false
        }
        for (i in 0..len-1) {
          if (squares[row][col+i] == '-') {
            squares[row][col+i] = 'b'
          } else {
            for (j in 0..i-1) {
              squares[row][col+j] = '-'
              return false
            }
          }
        }
      }
      false -> {
        if (row+len > 10) {
          return false
        }
        for (i in 0..len-1) {
          if (squares[row+i][col] == '-') {
            squares[row+i][col] = 'b'
          } else {
            for (j in 0..i-1) {
              squares[row+j][col] = '-'
            }
            return false
          }
        }
      }
    }
    return true
  }

  fun shoot(row:Int, col:Int):Int {
    when {
      row > 9 || col > 9 -> return -1
      squares[row][col] == 'm' || squares[row][col] == 'x' -> return 2
      squares[row][col] == '-' -> {
        squares[row][col] = 'm'
        return 0
      }
      squares[row][col] == 'b' -> {
        squares[row][col] = 'x'
        return 1
      }
    }
    return -1
  }

  fun gameOver():Boolean {
    for (row:Array<Char> in squares) {
      for (e:Char in row) {
        if (e == 'b') {
          return false
        }
      }
    }
    return true
  }

  fun foundShip(len:Int):Boolean {
    val tempSquares:Array<Array<Char>> = Array<Array<Char>>(10, {_ -> Array<Char>(10, {_ -> '-'})})
    for (i in 0..9) {
      for (j in 0..9) {
        tempSquares[i][j] = squares[i][j]
      }
    }
    for (i in 0..9) {
      for (j in 0..9) {
        if (tempSquares[i][j] == 'b' && j != 9 && tempSquares[i][j+1] == 'b') { // If square at [i][j] is a ship
          tempSquares[i][j] = '-' // Reset the sqaure so it doesn't trigger another event
          tempSquares[i][j+1] = '-'
          var found:Int = 2 // Already found 2 ships, looking for more
          for (k in j+2..9) { // For every element left in the row starting with 2 after j
            if (tempSquares[i][k] == 'b') { // If that square is also a ship
              tempSquares[i][k] = '-' // Reset the sqaure so it doesn't trigger another event
              found++ // Increment found
            } else { // If next square isn't a ship
              println("Found horiz ship, length $found")
              break // Break the loop
            }
          }
          if (found == len) { // After the loop completes, if the found ship matches specified length, return true
            return true
          }
        } else if (tempSquares[i][j] == 'b' && i != 9 && tempSquares[i+1][j] == 'b') { // If square below is also a ship
          tempSquares[i][j] = '-' // Reset the sqaure so it doesn't trigger another event
          tempSquares[i+1][j] = '-'
          var found:Int = 2 // Already found 2 ships, looking for more
          for (k in i+2..9) { // For every element left in the column starting with 2 after i
            if (tempSquares[k][j] == 'b') { // If that square is also a ship
              tempSquares[k][j] = '-' // Reset the sqaure so it doesn't trigger another event
              found++ // Increment found
            } else { // If next square isn't a ship
              println("Found vert ship, length $found")
              break // Break the loop
            }
          }
          if (found == len) { // After the loop completes, if the found ship matches specified length, return true
            return true
          }
        } else if (tempSquares[i][j] == 'b') { // Otherwise
          tempSquares[i][j] == '-' // Reset the square so it doesn't trigger another event
          if (len == 1) { // If length is 1
            return true
          }
        }
      }
    }
    return false // If all else fails, return false
  }
}